Unit 4 - Gaussian Splatting (#4)
huggingface/ml-for-3d-course: units/en/unit3/what-is-it.mdx

# What is it?

Gaussian Splatting is a **differentiable rasterization technique**.

## Differentiable Rasterization

In simple terms:

- Differentiable can be thought of as a fancy way to say "AI-compatible"
- Rasterization means taking data and drawing it on the screen

> Rasterization is already really common. Usually in the form of [triangle rasterization](https://en.wikipedia.org/wiki/Rasterisation), where 3D data is converted to 2D pixel data and drawn on the screen. That's how meshes are usually rendered.

Reviewer: ```suggestion
Rasterization is already really common. It usually takes the form of [triangle rasterization](https://en.wikipedia.org/wiki/Rasterisation), where 3D data is converted to 2D pixel data and drawn on the screen. That's how meshes are usually rendered.
```

![Mesh](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/ml-for-3d-course/mesh.png)

However, triangle rasterization isn't very AI-compatible. This is because it includes discrete decisions like:

> - Is this pixel inside the triangle?

Reviewer: Any others we could add?

> Neural networks don't like discrete decisions. They want everything to be fuzzy and continous - or in other words, *differentiable*.

Reviewer: ```suggestion
Neural networks don't like discrete decisions. They want everything to be fuzzy and continuous - or, in other words, *differentiable*.
```

## Gaussian Splatting

Gaussian Splatting is a differentiable rasterization technique. But how does it actually work?

Splats are composed of millions of points, where each point is composed of four parameters:

- **Position**: where it's located (XYZ)
- **Covariance**: how it's stretched (3x3 matrix)
- **Color**: what color it is (RGB)
- **Alpha**: how transparent it is (α)

> Then, to rasterize a splat, these points are projected into 2D. Then, for every pixel, contribute the contribution of every point. Or, in pseudocode:

Reviewer: ```suggestion
Then, to rasterize a splat, these points are projected into 2D. Then, for every pixel, add the contribution of every point. Or, in pseudocode:
```

> Then, to rasterize a splat, these points are projected into 2D. Then, for every pixel, contribute the contribution of every point. Or, in pseudocode:

Reviewer 2: I found the nomenclature `points` vs `pixels` a bit fuzzy. I understand that we start from a simplified conceptual framework, but maybe we could extend the "point" concept a little bit so we can introduce the concept of Gaussian?

```python

> splat2d = splat.project_and_sort()

Reviewer: Why is the sorting important if we end up adding all of them?

for point in splat2d:
    for pixel in image:
        pixel += compute_contribution(point, pixel)
```

> In theory, every point contribute to every pixel, which is very inefficient. However, that's okay, because it's *differentiable*.

Reviewer: ```suggestion
In theory, every point contributes to every pixel, which is very inefficient. However, that's okay because it's *differentiable*.
```

Reviewer 2: Why is it important that the process is differentiable in the context of rendering? And why is it ok that it is inefficient? Maybe clarify that it's ok for training, or simply remove the differentiable mention here?

> In practice, this is optimized with a tile-based rasterization method, as detailed in the [original paper](https://huggingface.co/papers/2308.04079).

Reviewer: Should people already know what is `tile-based rasterization`?

Reviewer 2: Maybe mention that the further away you move from a pixel position, the less important the contribution.

## Inference

> If you're not training a model, then it doesn't matter if it's differentiable. You can just treat each point as an instanced quad, as is in open-source web viewers like [gsplat.js](https://github.com/huggingface/gsplat.js).

Reviewer: ```suggestion
If you're not training a model, then it doesn't matter if it's differentiable. You can just treat each point as an instanced quad, as is in open-source web viewers like [gsplat.js](https://github.com/huggingface/gsplat.js).
```

> If you're not training a model, then it doesn't matter if it's differentiable. You can just treat each point as an instanced quad, as is in open-source web viewers like [gsplat.js](https://github.com/huggingface/gsplat.js).

Reviewer: What's a quad?

> If you're not training a model, then it doesn't matter if it's differentiable. You can just treat each point as an instanced quad, as is in open-source web viewers like [gsplat.js](https://github.com/huggingface/gsplat.js).

Reviewer 2: ```suggestion
If you're not training a model, then it doesn't matter if it's differentiable. You can just treat each point as an instanced quad, as in open-source web viewers like [gsplat.js](https://github.com/huggingface/gsplat.js).
```

> This can be seen in action [here](https://huggingface.co/spaces/dylanebert/igf).

Reviewer 2: Maybe explain what's going on there, or do you think it'd be repetitive?

## Training

> The [original paper](https://huggingface.co/papers/2308.04079) intializes the points using [Structure-from-Motion](https://en.wikipedia.org/wiki/Structure_from_motion), a traditional algorithm for 3D reconstruction.

Reviewer 2: ```suggestion
The [original paper](https://huggingface.co/papers/2308.04079) initializes the points using [Structure-from-Motion](https://en.wikipedia.org/wiki/Structure_from_motion), a traditional algorithm for 3D reconstruction.
```

![Structure from Motion](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/gaussian/points.png)

> These points are then rasterized using the tile-based method, and the loss is computed by comparing the rasterized image to the ground truth. Gradient descent is applies to adjust the point parameters (position, covariance, color, alpha).

Reviewer: ```suggestion
These points are then rasterized using the tile-based method, and the loss is computed by comparing the rasterized image to the ground truth. Gradient descent is applied to adjust the point parameters (position, covariance, color, alpha).
```

![Trained](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/gaussian/ellipsoids.png)

> The original paper also uses automated densification and pruning to automatically add and remove points as needed. More details can be found [here](https://huggingface.co/blog/gaussian-splatting).

Reviewer 2: I loved that post! Perhaps some of the gaussian plots could be reused here?

![Final](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/gaussian/bicycle.png)

## Generative 3D

> The original approach is suitable for learning individual scenes from photos. However, the concept is differentiable rasterization generalizes to more complex models like neural networks.

Reviewer 2: ```suggestion
The original approach is suitable for learning individual scenes from photos. However, the concept of differentiable rasterization generalizes to more complex models like neural networks.
```
